### DISCUSSION TITLE
Latency with 8 analog multiplexers (4067)

### DISCUSSION THREAD

**Discussion by gabooboo** [shanteacontrols/OpenDeck#90] [Q&A] [answered]
Hello again,

I am attempting to build a very large controller for use in Traktor.

I am using a custom board variant on an Arduino Mega 2560 with 8 4067 multiplexers for a total 128 analog inputs (split between rotary pots and analog buttons). In addition to the 8 multiplexers, I have 9 74HC595 shift registers connected giving a total of 72 single-color LEDs.

However, in both the web configurator and in Traktor, I am noticing a slight (~300ms) lag when moving the rotary pots.

I thought I might just be pushing past some limit of the Arduino, but then I noticed that when I set the analog input to "Button", this lag disappears.

This leads me to believe that the latency could instead be software related. Is there anything I can do about this issue?

Thanks in advance!

**Comment by paradajz** [comment, marked as answer]
Are you talking about initial latency, when you move the pot after it's been still for a while, or latency for each value?

**Reply by gabooboo** [reply]
After some careful testing, it does seem to be the initial latency.

If the pot has been still, the lag is there when I start to rotate the pot.

However, as I keep rotating it seems like the software and the pot then reach the upper/lower limits at the same time.

**Reply by paradajz** [reply]
Ah yes. ADC filter is currently tweaked primarily for STM32 boards. AVR has much slower readout but in most cases even that isn't an issue as long as the amount of analog inputs isn't large. This is not true in your case so I'm guessing the filtering can be simplified in this case. What you can try is to comment out lines 87-103 in `src/application/io/analog/Filter.h` like this:

![image](https://user-images.githubusercontent.com/2544094/125567145-658dcf8d-a3b2-426e-a959-3fbb50254ec8.png)

Hard to tweak these kind of things to work well for all architectures and all variants - impossible I'd say even.

**Reply by gabooboo** [reply]
Yep, that worked. The latency is gone for now.

Just curious, is there a board/architecture you recommend for the best performance?

Thank you again!

**Reply by paradajz** [reply]
I knew it would get rid of latency, but what I'm more interested in is the stability of the readouts. Has anything changed regarding that? If you're not getting jumpy values from time to time then I'll disable the initial filtering on ADC when the component count is larger than some number on AVR boards. Probably 32.

The architecture should be irrelevant to users. Yes, the STM32 is faster and offers higher resolution ADC and room for more presets, but other than that things are pretty much identical, so Arduino Mega should be just fine.

STM32 boards are generally simpler to flash so there's that as well (especially Black Pill boards).

**Reply by gabooboo** [reply]
I didn't notice any "jumping" of values yet. The pots are certainly more sensitive. Even rotating a pot within it's deadzone changes the value +- 1 or 2. But so far I have no issue with that.

I'm still connecting everything and testing so I will report back with more info as soon as possible.

And thanks for the recommendation. I will look into STM32 boards as I've never used them before.

**Comment by paradajz** [comment]
For anyone stumbling on this: analog filters can now be disabled via YAML config.
https://github.com/shanteacontrols/OpenDeck/wiki/Creating-custom-board-variant#analog
